### ISSUE TITLE
Missing functions in shared librarry 

### ISSUE THREAD

**Issue by gidesa** [svaarala/duktape#960]
In the Duktape shared library (.dll, .so) are missing some  functions reported in API guide.
These functions are very useful to a better comunication with the embedding program (C program, or whatever):

duk_peval_string
duk_eval
duk_error
duk_push_error_object

**Comment by harold-b** [comment]
Those are defines as preprocessor macros so they wont be compiled into any object files. You either have to include duktape.h, of if you're loading the functions dynamically (which I assume is why your running into this issue) you'll have to copy over at least those macros into your project.

**Comment by svaarala** [comment]
You should always include `duktape.h`: otherwise the compiler will be guessing the types involved which won't always work.

**Comment by gidesa** [comment]
Thanks for your answers. Indeed I am using the Duktape shared library calling it from a Delphi Pascal project, so I cannot include duktape.h in the caller, or use directly the C macros.
The problem with macros is that they use filename and line number  using  standard C ambient variables (__LINE__ , __FILE__). Of course they are inaccessible from any external program calling the shared library. So I cannot simply rewrite them inside the caller.

**Comment by svaarala** [comment]
@gidesa This may be a duplicate to #87?

Re: `__FILE__` and `__LINE__` the intent is to ensure that the API provides variants with explicit file/line arguments for each such call so that the file/line "blaming" can be controlled by the application - #95. This is not yet the case however.

**Comment by gidesa** [comment]
Yes, my question is a duplicate of #87 in some form, although it's more specific. I don't think that in shared library the macros are a serious problem in itself, because one could always rewrite them in the external language (in my case as functions/procedures in Delphi Pascal). The problem are variables internal to Duktape, or to C compiler, not exposed  to external language.
Maybe there are other ways for caller to read the current file name and line of the script executed by Duktape. Or maybe  future versions could add a specific API function for this task.

**Comment by svaarala** [comment]
@gidesa Hmm. The calls you mention use a macro to automatically supply `__FILE__` and `__LINE__` to Duktape internals. For example `duk_error()` provides them so that the file/line information is recorded in the stack trace. But if a call was added to allow supplying these explicitly, say, `duk_error_fileline(ctx, "myfile.xxx", 123, "connect failed")`, would that solve your problem? (That's the goal of #95 and there are really only a few calls that lack explicit file/line argument variants.)

> Maybe there are other ways for caller to read the current file name and line of the script executed by Duktape.

Do you mean here that when a Duktape/C function is called, you would like to find out the file/line of the *calling* function? There's some support for that via `Duktape.act()` but you're right in that there's no easy-to-use API for that specific purpose for now. Definitely an area of improvement.

> Embedding it it's especially simple in Delphi, only a few lines of interfacing code and you have a complete JS interpreter ready inside your program.

Good to hear things are still working in your environment :)

**Comment by gidesa** [comment]
> Do you mean here that when a Duktape/C function is called, you would like to find out the file/line of the calling function? There's some support for that via Duktape.act() but you're right in that there's no easy-to-use API for that specific purpose for now. Definitely an area of improvement. 

Yes, in some respect. Consider an ideal situation: the Duk interpreter at some point calls a function in the host environment (Delphi). Then here something goes wrong,  and I would return my custom error message and code to Duk interpreter. Then Duk could decide to print the error, with message and error code and so on, adding the file name and line number, but instead could also decide to "eat" entirely the error, if a catch is present.
Actually I resolve to generate an error in Duk returning -1 from the called (host) function, so I can have the filename/line printed from standard Duk error message. But the host error message can only be directly printed to output in the called function, before it returns to caller Duk function. This is sufficient in many cases, but if a catch is present the host error message is printed anyway, because it's not managed by Duk interpreter at all. 
So the duk_error macro does exactly that, pass a custom error message and code to Duk, and it decides what to do.

**Comment by gidesa** [comment]
I have to complete the comment: if duk_error is a macro, then not exposed in shared library, one could rewrite them in the external (host) language, but is needed a real function in Duk API (exposed in library)  to read the current JS file name and line number.

**Comment by svaarala** [comment]
@gidesa Why would current JS file name and line number be needed? Note that `duk_error()` includes the *C* `__FILE__` and `__LINE__` information to the stack trace data - it's not related to JS file/line information. Any Ecmascript functions (functions within JS files) will already be included the stack trace data and their file/line information will be based on the bytecode executor state for each such function.

But maybe what you're after is that when you access `.fileName` and `.lineNumber` of an error object, those would reflect the Delphi call site? Having the ability to provide file/line explicitly would do that.

In some cases one would actually want to *ignore* the C call site file/line when assigning .fileName / .lineNumber (but they would still be included in a stack trace). See links in #458.

**Comment by gidesa** [comment]
@svaarala  Ok, I understand that `__LINE__ and __FILE__` are not referenced to the JS script.
I just need the file/line of JS script, but they are just in some form in the stack trace. I have to see more on this, in the documentation.
Yes, I need the .fileName and .lineNumber directly in the called host function.  For now I can read them only after the end of evaluation ( I use  duk_eval_raw). If the host function return -1, Duk generate an error object with message "Unknown error" and JS filename and line. So I read them with: 
          duk_get_prop_string(ctx, -1, 'fileName');
          duk_get_prop_string(ctx, -2, 'lineNumber');
          duk_get_prop_string(ctx, -3, 'message');
Instead of the "Unknown error" message I would have the specific message from the host function. But I don't know how to pass that message to Duk evaluator.
So indeed there are two solution to my little problem:
- read JS filename and number inside the host function, and print directly the complete message
or
- pass an entire error message string from host function to Duk evaluator, other than returning the -1 code (better because evaluator can decide to stop or continue processing, based on try-catch blocks).
